Supporting FIPS work

mshah0722 opened the pull request:
- [ ] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

################################################################################################

# Delete this section and fill in the remaining info from the template

ATTENTION, READ THIS: Updated July 2024 

Read and understand this completely, then delete the static part of the template. 

If a reviewer or merger sees this template, they should fail the review or merge.

If this code change is fixing a user-visible bug in previously released code, it MUST
have an associated issue **mentioned in the PR text or description**.  
 
   - That Issue also MUST be labelled “release bug”

     This directs automation to scrape this fix for inclusion in the next release's
     list of bugs fixed.

   - The linkage between PR and Issue should use the `Fixes #...` syntax rather than a manual
     link via the Development widget.

Otherwise, a link to an issue or specific issue labels are optional.

For full details, please see this wiki page: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions
################################################################################################

utle committed:
Support FIPS140-3

utle committed:
clean up

mshah0722 committed:
Supporting FIPS work

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

mshah0722 committed:
Minor bug fixes

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

mshah0722 committed:
Minor changes

utle committed:
Use the common crypto bundle

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

utle committed:
Create the Provider class

utle committed:
clean up

mshah0722 committed:
Adding support to client and server for java 17

utle committed:
Rename generate3DESKey method to generateSharedKey

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

!platform-semeru17
jvmUnderTestPath=/liberty/jvms/linux_x86_64/ibm-semerujdk-17.0.12+7_openj9-0.46.1/jdk-17.0.12+7.tar.gz

mshah0722 commented:
!customSOE
artifact_execution_id=_gMTjAHdfEe-ZBdXWdK7Miw
platforms=CENTOS9_X86_64_BETA_FIPS_IBM_8.yml,WINDOWS10_X86_64_BETA_FIPS_IBM_8.yml,UBUNTU20_PPC64LE_BETA_FIPS_IBM_8.yml
jvmUnderTestPath=/liberty/jvms/linux_x86_64/ibm-semerujdk-21.0.4+7_openj9-0.46.1/jdk-21.0.4+7.tar.gz

mshah0722 commented:
!customSOE
artifact_execution_id=_gMTjAHdfEe-ZBdXWdK7Miw
platforms=CENTOS9_X86_64_BETA_FIPS_IBM_8.yml,WINDOWS10_X86_64_BETA_FIPS_IBM_8.yml,UBUNTU20_PPC64LE_BETA_FIPS_IBM_8.yml
jvmUnderTestPath=/liberty/jvms/linux_x86_64/ibm-semerujdk-17.0.12+7_openj9-0.46.1/jdk-17.0.12+7.tar.gz

mshah0722 commented:
#libby

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Fixing some test bugs

mshah0722 committed:
Merge branch 'FIPSwork' of github.com:mshah0722/open-liberty into HEAD

mshah0722 committed:
Merge pull request #1 from utle/FIPSwork

Rename generate3DESKey method to generateSharedKey

mshah0722 committed:
Minor bugs

utle committed:
Added debug msgs

utle committed:
SSL to use crypto common code

mshah0722 committed:
Merge pull request #2 from utle/FIPSwork

Fips work

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Restrict TLS versions for FIPS

mshah0722 committed:
Merge branch 'FIPSwork' of github.com:mshah0722/open-liberty into FIPSwork

mshah0722 committed:
Restrict TLS versions for FIPS

utle committed:
Handle Semeru

mshah0722 committed:
Merging Ut's code

utle committed:
add common crypto to ssl feature

mshah0722 committed:
Merge pull request #4 from utle/FIPSwork

add common crypto to ssl feature

utle committed:
enable openJCEPlusFips

utle committed:
enable openJCEPlusFips

utle committed:
clean up

mshah0722 committed:
Merge pull request #5 from utle/FIPSwork

Fip swork

mshah0722 committed:
Merge remote-tracking branch 'ut_branch/FIPSwork' into FIPSwork

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Unit Test fixes

mshah0722 committed:
Creating a new key when fips is enabled for testing

mshah0722 committed:
Creating a new key when fips is enabled for testing

utle committed:
Create and place the CryptoUtils in the kernel

mshah0722 committed:
Merge pull request #6 from utle/FIPSwork

Create and place the CryptoUtils in the kernel

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

utle committed:
audit to use the common crypto code

utle committed:
missing package rename

mshah0722 committed:
Merge pull request #7 from utle/FIPSwork

missing package rename

mshah0722 committed:
FIPS enabled FAT fixes

utle committed:
get the right jdk provider

utle committed:
Refactor encrypt and decrypt methods

utle committed:
add trivial

utle committed:
Increase log max allow

utle committed:
import the missing package

utle committed:
testing

utle committed:
import the missing package

mshah0722 committed:
Fixing pre-generated keys

mshah0722 committed:
Fixing FAT Failures

mshah0722 committed:
Fixing FAT Failures

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Merge pull request #8 from utle/FIPSwork

testing

mshah0722 committed:
Fixing FAT failures

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Merge pull request #9 from utle/FIPSwork

import the missing package

mshah0722 committed:
Fixing FAT issues

mshah0722 committed:
Fixing FAT errors

mshah0722 committed:
Minor bug with loading pre-generated keys

mshah0722 committed:
FAT fixes

mshah0722 committed:
Merge branch 'FIPSwork' into FIPSwork

mshah0722 committed:
Merge pull request #10 from utle/FIPSwork

import the missing package

mshah0722 committed:
FAT fixes

utle committed:
Removve com.ibm.ws.common.crypto

utle committed:
Clean up

utle committed:
comments out the @Trivial

utle committed:
add debug msg

mshah0722 committed:
Merge branch 'FIPSwork' into FIPSwork

mshah0722 committed:
Merge pull request #11 from utle/FIPSwork

Removve com.ibm.ws.common.crypto

mshah0722 committed:
Fixing FAT Failures

mshah0722 committed:
Adding the last few missing FIPS keys

utle committed:
Remove comments code

mshah0722 committed:
Going through LTPA FAT Failures

mshah0722 committed:
Updating leftover buildpaths

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

utle committed:
Using RSA algorithm

mshah0722 committed:
Updating leftover buildpaths

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Fixing FAT Failures

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Compile error

mshah0722 committed:
Merge pull request #12 from utle/FIPSwork-audit

Fips work audit

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

utle committed:
Remove check for the provider class available

mshah0722 committed:
Merge pull request #13 from utle/FIPSwork-audit

Remove check for the provider class available

utle committed:
Move messageDigest getInstance to CryptoUtil

mshah0722 committed:
Merge branch 'FIPSwork' into FIPSwork-audit

mshah0722 committed:
Merge pull request #14 from utle/FIPSwork-audit

Move messageDigest getInstance to CryptoUtil

mshah0722 committed:
Resolving BVT issue

mshah0722 committed:
Testing new LibertyServer code

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Remove duplicate change

mshah0722 committed:
Minor modifications

mshah0722 committed:
Reading ltpaFIPS keys in fattest simplicity

utle committed:
Undo changes for LogPolice

mshah0722 committed:
Merge pull request #15 from utle/FIPSwork

Undo changes for LogPolice

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Avoid null issue in message digest initialization

mshah0722 committed:
Working through FAT Failures

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 committed:
Working through FAT Failures

Deleting ltpa.keys when FIPS keys not found

mshah0722 committed:
Merge branch 'OpenLiberty:integration' into FIPSwork

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

mshah0722 committed:
Fixing md1 null issue

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

mshah0722 committed:
Removing whiteshpace changes

mshah0722 commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

jhanders34 commented on dev/com.ibm.ws.kernel.service/src/com/ibm/ws/common/crypto/CryptoUtils.java:
```diff
@@ -0,0 +1,382 @@
+                    Tr.error(tc, "FIPS is enabled but the " + IBMJCE_PLUS_FIPS_PROVIDER + " provider is not available.");
+                }
+                ibmJCEPlusFIPSAvailable = false;
+            }
```
```suggestion
            boolean ibmJCEPlusFIPSProviderAvailable = JavaInfo.isSystemClassAvailable(IBMJCE_PLUS_FIPS_PROVIDER);
            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                Tr.debug(tc, "ibmJCEPlusFIPSProvider: " + IBMJCE_PLUS_FIPS_PROVIDER);
                Tr.debug(tc, "ibmJCEPlusFIPSProviderAvailable: " + ibmJCEPlusFIPSProviderAvailable);
            }

            if (ibmJCEPlusFIPSProviderAvailable) {
                ibmJCEPlusFIPSAvailable = isRunningBetaMode();
            } else {
                if (fipsEnabled && !isSemeruFips()) {
                    Tr.error(tc, "FIPS is enabled but the " + IBMJCE_PLUS_FIPS_PROVIDER + " provider is not available.");
                }
                ibmJCEPlusFIPSAvailable = false;
            }
            ibmJCEPlusFIPSProviderChecked = true;
```

jhanders34 commented on dev/com.ibm.ws.kernel.service/src/com/ibm/ws/common/crypto/CryptoUtils.java:
```diff
@@ -0,0 +1,382 @@
+                ibmJCEPlusFIPSAvailable = true;
+            } else {
+                if (fipsEnabled && !isSemeruFips()) {
+                    Tr.error(tc, "FIPS is enabled but the " + IBMJCE_PLUS_FIPS_PROVIDER + " provider is not available.");
```
Tr.error requires a translated message.  Is this temporary?

utle replied:
This is temporary for now.

jhanders34 commented on dev/com.ibm.ws.kernel.service/src/com/ibm/ws/common/crypto/CryptoUtils.java:
```diff
@@ -0,0 +1,382 @@
+        if (openJCEPlusFIPSProviderChecked) {
+            return openJCEPlusFIPSAvailable;
+        } else {
+            openJCEPlusFIPSAvailable = JavaInfo.isSystemClassAvailable(OPENJCE_PLUS_FIPS_PROVIDER);
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "openJCEPlusFIPSProvider: " + OPENJCE_PLUS_FIPS_PROVIDER);
+            }
+
+            openJCEPlusFIPSProviderChecked = true;
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "openJCEPlusFIPSAvailable: " + openJCEPlusFIPSAvailable);
+            }
+
+            if (isRunningBetaMode() && openJCEPlusFIPSAvailable) {
+                openJCEPlusFIPSAvailable = true;
+            } else {
+                if (fipsEnabled && isSemeruFips()) {
+                    Tr.error(tc, "Semeru FIPS is enabled but the " + OPENJCE_PLUS_FIPS_PROVIDER + " provider is not available.");
+                }
+                openJCEPlusFIPSAvailable = false;
+            }
```
```suggestion
            boolean openJCEPlusFIPSProviderAvailable = JavaInfo.isSystemClassAvailable(OPENJCE_PLUS_FIPS_PROVIDER);
            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                Tr.debug(tc, "openJCEPlusFIPSProvider: " + OPENJCE_PLUS_FIPS_PROVIDER);
                Tr.debug(tc, "openJCEPlusFIPSAvailable: " + openJCEPlusFIPSProviderAvailable);
            }

            if (openJCEPlusFIPSProviderAvailable) {
                openJCEPlusFIPSAvailable = isRunningBetaMode();
            } else {
                if (fipsEnabled && isSemeruFips()) {
                    Tr.error(tc, "Semeru FIPS is enabled but the " + OPENJCE_PLUS_FIPS_PROVIDER + " provider is not available.");
                }
                openJCEPlusFIPSAvailable = false;
            }
            openJCEPlusFIPSProviderChecked = true;
```

jhanders34 commented on dev/com.ibm.ws.kernel.service/src/com/ibm/ws/common/crypto/CryptoUtils.java:
```diff
@@ -0,0 +1,382 @@
+                openJCEPlusFIPSAvailable = true;
+            } else {
+                if (fipsEnabled && isSemeruFips()) {
+                    Tr.error(tc, "Semeru FIPS is enabled but the " + OPENJCE_PLUS_FIPS_PROVIDER + " provider is not available.");
```
Similar statement about Tr.error requiring a translated message.

utle replied:
This is temporary for now.

jhanders34 commented on dev/com.ibm.ws.kernel.service/src/com/ibm/ws/common/crypto/CryptoUtils.java:
```diff
@@ -0,0 +1,382 @@
+        } else {
+            // Running beta exception, issue message if we haven't already issued one for this class
+            if (!issuedBetaMessage) {
+                Tr.info(tc, "BETA: A beta method has been invoked for the class CryptoUtils for the first time.");
```
Today this message will come out even if the user has not configured 140-3 for example in `isFips140_3Enabled()` because the conditional is backwards in there.  It check to see if beta it set before checking to see if 140-3 is set.  No reason to put out a message unless they are using function that requires beta.

utle replied:
Fixed it.

jhanders34 commented on dev/com.ibm.ws.kernel.service/src/com/ibm/ws/common/crypto/CryptoUtils.java:
```diff
@@ -0,0 +1,382 @@
+    }
+
+    public static boolean isFips140_2Enabled() {
+        return isRunningBetaMode() && "140-2".equals(FIPSLevel);
```
Don't we already have 140-2 support?  Won't this break that?

utle replied:
140-2 is baked in with the code ... we don't have a flag for 140-2 yet. ...

utle committed:
Fixed code review comments

mshah0722 committed:
Merge branch 'release' of github.com:OpenLiberty/open-liberty into FIPSwork

mshah0722 commented:
#libby

!build
git.clone.checkout.branch=FIPSwork
git.clone.repository.username=Malhar-Shah

#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

jimmy1wu committed:
add com.ibm.ws.common.crypto trace

jimmy1wu committed:
add com.ibm.ws.common.crypto trace

jimmy1wu committed:
add com.ibm.ws.common.crypto trace

jimmy1wu committed:
add com.ibm.ws.common.crypto trace

jimmy1wu committed:
add com.ibm.ws.common.crypto trace

jimmy1wu committed:
fix up some traces

jimmy1wu committed:
Merge pull request #17 from jimmy1wu/FIPSwork

add com.ibm.ws.common.crypto trace

jacobwdv committed:
Updating LibertyServer class so that other tests can configure LTPA keys and JVM options for FIPS (#16)

* updating LibertyServer

mshah0722 committed:
Merge branch 'release' of github.com:OpenLiberty/open-liberty into FIPSwork

mshah0722 commented:
#libby 
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=9c07e88e-ee4a-44f8-a893-b57f28c176e6) - :x: completed with errors/failures.) ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6badf821-3f1f-4574-a6ad-54eaa1a534aa) - :white_check_mark: completed successfully!)

!build
git.clone.checkout.branch=FIPSwork
git.clone.repository.username=Malhar-Shah

#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

mshah0722 commented:
#libby 
 ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=da5602e1-da2b-4ac2-bcb1-496ca04f98b9) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=5c213b91-8595-4a6d-a840-973a899eaeac) - :x: completed with errors/failures.)

!build
git.clone.checkout.branch=FIPSwork
git.clone.repository.username=Malhar-Shah

#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

mshah0722 committed:
Fixing SSL and beta mode issue for Java 8

mshah0722 commented:
#libby
 ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=68e0b6e2-91c9-4070-be1a-93f909669578) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=7e744a04-e253-4081-bda6-1e68b982ac3b) - :x: completed with errors/failures.)

!build
git.clone.checkout.branch=FIPSwork
git.clone.repository.username=Malhar-Shah

#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

mshah0722 committed:
Fixed the issue with IBMJCEPlusFIPSprovider not loading properly

mshah0722 committed:
Merge branch 'release' of github.com:OpenLiberty/open-liberty into FIPSwork

mshah0722 commented:
#libby
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=a5ee5152-23e3-4e5a-9605-a507ddd78274) - :x: completed with errors/failures.) ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=069337bb-23a1-42a5-bb0d-0dda0dd40019) - :white_check_mark: completed successfully!)

!build
git.clone.checkout.branch=FIPSwork
git.clone.repository.username=Malhar-Shah

#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

jacobwdv committed:
updating server xml's with appropriate passwords, renaming serverFIPS.xml before server start

adding AEADBadTagException to AllowedFFDC list for tests which expect BadPaddingException

using jvm.options so that mbean invocations will use proper jvm parameters

utle committed:
Merge branch 'FIPSwork' of github.com:mshah0722/open-liberty into FIPSwork

utle committed:
Fixed 140-2 check and adding more debug

utle committed:
adding more check

utle committed:
adding more check

mshah0722 committed:
Merge branch 'release' of github.com:OpenLiberty/open-liberty into FIPSwork

mshah0722 committed:
Merge branch 'FIPSwork' of github.com:utle/open-liberty into FIPSwork

mshah0722 committed:
Merge branch 'release' of github.com:OpenLiberty/open-liberty into FIPSwork

mshah0722 commented:
#libby
 ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=99e4bc55-d537-44d5-84a0-41cb68e3ca52) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=086b03d5-74f9-4585-ad56-8fce553b61c9) - :x: completed with errors/failures.)

!build
git.clone.checkout.branch=FIPSwork
git.clone.repository.username=Malhar-Shah

#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

jacobwdv committed:
wip

mshah0722 committed:
Merge branch 'release' of github.com:OpenLiberty/open-liberty into FIPSwork

mshah0722 committed:
Merge pull request #18 from jacobwdv/JD_LTPAKeyTest

updating server xml's with appropriate passwords, renaming serverFIPS…

mshah0722 committed:
Merge branch 'FIPSwork' of github.com:mshah0722/open-liberty into FIPSwork

mshah0722 commented:
#libby
 ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=fe797001-01af-4954-a931-2cc464e3a416) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=9d27789e-4d64-427e-85ef-53e15b7c8bb4) - :x: completed with errors/failures.)

!build
git.clone.checkout.branch=FIPSwork
git.clone.repository.username=Malhar-Shah

#spawn.fullfat.buckets=com.ibm.ws.security.token.ltpa_fat

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

mshah0722 committed:
Merge branch 'release' of github.com:OpenLiberty/open-liberty into FIPSwork

mshah0722 committed:
Test fixing nonFIPS ltpa fat

jhanders34 commented:
#30502 / #30625 were merged in now so I believe this can be closed.
